bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/_roodataset.py

```diff
@@ -54,3 +54,200 @@ def plotOnXY(self, *args, **kwargs):
+            arr = arr if arr.dtype == np.int32 else np.array(arr, dtype=np.int32)
+            vec.resize(len(arr))
+
+            beg = arr.ctypes.data_as(ctypes.POINTER(ctypes.c_int))
+            void_p = ctypes.cast(beg, ctypes.c_voidp).value + 4 * len(arr)
+            end = ctypes.cast(void_p, ctypes.POINTER(ctypes.c_int))
+            ROOT.std.copy(beg, end, vec.begin())
+
+        dataset.store().recomputeSumWeight()
+
+        n_entries = None
+
+        for real in dataset.store().realStoreList():
+            if n_entries is None:
+                n_entries = real.size()
+            assert n_entries == real.size()
+
+        for cat in dataset.store().catStoreList():
+            assert n_entries == cat.size()
+
+        return dataset
+
+    def to_numpy(self, copy=True, compute_derived_weight=False):
+        """Export a RooDataSet to a dictinary of numpy arrays.
+
+        Args:
+            copy (bool): If False, the data will not be copied. Use with
+                         caution, as the numpy arrays and the RooAbsData now
+                         own the same memory. If the dataset uses a
+                         RooTreeDataStore, there will always be a copy and the
+                         copy argument is ignored.
+            compute_derived_weight (bool): Sometimes, the weight variable is
+                not stored in the dataset, but it is a derived variable like a
+                RooFormulaVar. If the compute_derived_weight is True, the
+                weights will be computed in this case and also stored in the
+                output. Switched off by default because the computation is
+                relatively expensive.
+
+        Returns:
+            dict: A dictionary with the variable or weight names as keys and
+                  the numpy arrays as values.
+        """
+        import ROOT
+        import numpy as np
+
+        data = {}
+
+        if isinstance(self.store(), ROOT.RooVectorDataStore):
+            for name, array in self.store().to_numpy(copy=copy).items():
+                data[name] = array
+        elif isinstance(self.store(), ROOT.RooTreeDataStore):
+            # first create a VectorDataStore so we can read arrays
+            store = self.store()
+            variables = store.get()
+            store_name = store.GetName()
+            tmp_store = ROOT.RooVectorDataStore(store, variables, store_name)
+            for name, array in tmp_store.to_numpy(copy=copy).items():
+                data[name] = array
+        else:
+            raise RuntimeError(
+                "Exporting RooDataSet to numpy arrays failed. The data store type "
+                + self.store().__class__.__name__
+                + " is not supported."
+            )
+
+        # Special case where the weight is a derived variable (e.g. a RooFormulaVar).
+        # We don't want to miss putting the weight in the output arrays, so we
+        # are forced to iterate over the dataset to compute the weight.
```

Comment:
Can we perform the weight computation ny iteration on the data set in C++ and just call that function from Python ?

Reply:
I remember now why I didn't do this in C++! The `RooAbsData::get` function is very slow, especially if it triggers the reevaluation of the computation graph for the weight variable like in this case. I assumed the overhead from this will be larger than the Python overhead here. This is something that should be verified in the future!